src/analysis/score.rs

```diff
@@ -24,6 +30,94 @@ pub fn row_accuracy(outputs: &Matrix<f64>, targets: &Matrix<f64>) -> f64 {
     accuracy(outputs.iter_rows(), targets.iter_rows())
 }
 
+/// Returns the precision score for 2 class classification.
+/// true-positive / (true-positive + false-positive)
+pub fn precision<'a, I, T>(outputs: I, targets: I) -> f64
+    where I: ExactSizeIterator<Item=&'a T>,
+          T: 'a + PartialEq + Zero + One
+{
+    assert!(outputs.len() == targets.len());
+
+    let mut tpfp = 0.0f64;
+    let mut tp = 0.0f64;
+
+    for (ref o, ref t) in outputs.zip(targets) {
+        if *o == &T::one() {
+            tpfp += 1.0f64;
+            if *t == &T::one() {
+                tp += 1.0f64;
+            }
+        }
+        if ((*t != &T::zero()) & (*t != &T::one())) |
+           ((*o != &T::zero()) & (*o != &T::one())) {
+            panic!("precision must be used for 2 class classification")
+        }
+    }
+    tp / tpfp
+}
+
+/// Returns the recall score for 2 class classification.
+/// true-positive / (true-positive + false-negative)
+pub fn recall<'a, I, T>(outputs: I, targets: I) -> f64
+    where I: ExactSizeIterator<Item=&'a T>,
+          T: 'a + PartialEq + Zero + One
+{
+    assert!(outputs.len() == targets.len());
+
+    let mut tpfn = 0.0f64;
+    let mut tp = 0.0f64;
+
+    for (ref o, ref t) in outputs.zip(targets) {
+        if *t == &T::one() {
+            tpfn += 1.0f64;
+            if *o == &T::one() {
+                tp += 1.0f64;
+            }
+        }
+        if ((*t != &T::zero()) & (*t != &T::one())) |
+           ((*o != &T::zero()) & (*o != &T::one())) {
+            panic!("recall must be used for 2 class classification")
+        }
+    }
+    tp / tpfn
+}
+
+/// Returns the f1 score for 2 class classification
+/// 2 * precision * recall / (precision + recall)
+pub fn f1<'a, I, T>(outputs: I, targets: I) -> f64
+    where I: ExactSizeIterator<Item=&'a T>,
+          T: 'a + PartialEq + Zero + One
+{
+    assert!(outputs.len() == targets.len());
+
+    let mut tpos = 0.0f64;
+    let mut fpos = 0.0f64;
+    let mut fneg = 0.0f64;
+
+    for (ref o, ref t) in outputs.zip(targets) {
+        if (*o == &T::one()) & (*t == &T::one()) {
+            tpos += 1.0f64;
+        } else if *t == &T::one() {
+            fpos += 1.0f64;
+        } else if *o == &T::one() {
+            fneg += 1.0f64;
+        }
+        if ((*t != &T::zero()) & (*t != &T::one())) |
+           ((*o != &T::zero()) & (*o != &T::one())) {
+            panic!("f1-score must be used for 2 class classification")
+        }
+    }
+    // precision
+    let p = tpos / (tpos + fpos);
```

Comment:
We can simplify this a little, [see this article](https://en.wikipedia.org/wiki/F1_score). For the f1 score:

```rust
2 * tpos / (2 * tpos + fneg + fpos)
```